src/v/strings/static_str.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include <fmt/format.h>
+
+#include <string_view>
+#include <type_traits>
+
+#pragma once
+
+/**
+ * @brief A wrapper around a compile-time constant string.
+ *
+ * A static_str wraps an compile-time constant string, and is intended to be
+ * used at runtime as a guarantee that a string has static lifetime. Such string
+ * have trivial and cheap copy operations, since we just copy pointers (or
+ * views) to the underlying string.
+ */
+class static_str {
+public:
+    /**
+     * @brief Create a static_str based on a compile-time literal.
+     *
+     * The use of consteval prevents passing anything other than compile-time
+     * values, which necessarily have static lifetime.
+     */
+    consteval static_str(const char* str) // NOLINT
```

Comment:
this is fairly niche, but playing around a bit in godbolt, if this consteval constructor takes `std::string_view` then it seems to still works as expected accepts a string literal. but when taking `const char *` it won't accept a constant expression std::string_view like this

```
consteval int a(const char *x) {
    return std::string_view(x).size();
}

consteval std::string_view b() {
    return {};
}

int main() {
    return a(b());
}
```

Reply:
Oh yeah like the idea of taking a std::string_view, I seems 'strictly better' in a way, but then things like this fail to compile:

```
void takes_static_str(static_str s) {
    // ...
}

void send_std_view() {
    takes_static_str("abc");
}
```

Probably because two conversions are required. Passing string literals are kind of the core use-case I am imagining, so it should be low friction. 

However, I can simply offer _both_ constructors to make this work.

Reply:
Both constructors added https://github.com/redpanda-data/redpanda/commit/945ef20cca67bfdf84a23d621084cea2ee48f084.

Reply:
Both constructors added in https://github.com/redpanda-data/redpanda/commit/945ef20cca67bfdf84a23d621084cea2ee48f084